Siddarth Rayapur

Data Mining 700

3/15/2026

# Dataset Selection (checkpoint 1 RECAP)

**Online Retail Dataset**

Reasons:


*   Clearly supports frequent itemset and association rule mining (course).
*   Transaction timestamps allow for squential pattern mining (beyond the course).
*   Dataset is large enough to find meaningful patterns and analyze.
*   Inclusion of real-world issues like returns and skewed item frequencies.

Tradeoffs:


*   Does not have any information on the demographic.
*   no prediction target.





# Explaratory Data Analysis

This section analyzed the structure, quality, and characteristics of the online retail dataset before data mining takes place.

In [ ]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

online_retail = pd.read_excel("/content/drive/MyDrive/colab_notebooks/Data_mining/Online Retail.xlsx")
online_retail.head()

Mounted at /content/drive


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [ ]:
online_retail.shape

(541909, 8)

In [ ]:
online_retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [ ]:
online_retail.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


This dataset has 541909 transaction records and 8 variables. Each row represents a product within a transaction. Most of the variables are complete, except for Description and CustomerID, which has missing values. There are alos negative values in the Quantity variable and UnitPrice. This implies that some of the items could have been returned. This suggests a need for data cleaning before conducting the itemset mining. The span of the dataset is from December 2010 to December 2011 and has Transactional, Temporal, and information on pricing, making it a good for pattern mining analysis.

In [ ]:
# remove cancelled invoices
online_retail = online_retail[~online_retail['InvoiceNo'].astype(str).str.startswith('C')]

# remove -/0
online_retail = online_retail[online_retail['Quantity'] > 0]

# remove -/0 prices
online_retail = online_retail[online_retail['UnitPrice'] > 0]

online_retail.shape


(530104, 8)

After removing the cancelled invoices, negative values, and invalid prices, the dataset was reduced 530104 rows. This ensures that only completed purchases are analyzed.

In [ ]:
basket_size = online_retail.groupby('InvoiceNo')['StockCode'].count()
print(basket_size.describe())

count    19960.000000
mean        26.558317
std         47.542066
min          1.000000
25%          6.000000
50%         15.000000
75%         29.000000
max       1114.000000
Name: StockCode, dtype: float64


## Checkpoint 2 - Research Questions Formation

# Project Scope
For this project, Online Retail dataset was used. This dataset contains transactional records from a United Kingdom based online retailer. The Online Retail dataset is ideal for basket-based analysis because each of the transactions include several products.

<br> From the EDA we have significant patterns that were observed. Most of the transactions contain a small number of items and a few of the transactions contain large quantities, causing the basket sizes to be highly skewed. The product frequency also shows a long-tail distribution, in which a small nummber of the products appear often while most of the products appear rarely. last but not least, the transaction volume indicates temporal fluctuation, with signs of it increasing towards the end of the year.

<br> These characteristics show that the Online Retail dataset is suitable for frequent itemset mining and association rule mining (course techniques). With the presence of timestamp data, sequential pattern (external technique) mining can be used as well.



# Research Question Definition
RQ1: What frequent itemset appear in the Online Retail dataset when varying the support threshold?
* *Data mining task type:* Frequent itemset mining
* *Relevant algorithms:* Apriori (course), FP-Growth
* *Evaluation Criteria:* Support and number of itemset

<br> RQ2: How do the confidence and lift compare when dealing with association rules evaluated from the retail transactions?
* *Data mining task type:* Association Rule mining
* *Relevant algorithms:* Association Rule generation
* *Evaluation Criteria:* Confidence, lift, and support

<br> RQ3: Can Sequntial patterns show purchasing relationships that would not be captured by unordered frequent itemsets?
* *Data mining task type:* sequential pattern mining
* *Relevant algorithms:* PrefixSpan (external)
* *Evaluation Criteria:* Pattern support and interpretability

# RQ-to-Method Mapping
RQ1 focuses on frequent itemset mining and will find products that appear more often using methods like Apriori and FP-Growth. The evaluation metric will be support.

<br> RQ2 focuses on association rule mining. This will check the strength and usefulness of the rules received from frequent itemsets using confidence and lift.

<br> RQ3 uses an algorithm like PrefixSpan for the sequential pattern mining. The purpose is to determine whether having the transaction order shows any new patterns that are not captured by frequent itemset mining.

# Motivation and Feasibility
The EDA showed many significant characteristics of the dataset, such as long-tail distribution of products, large variation in basket sizes, and clear temporal patterns in the transaction activity. These results encourage the usage of both techniques learned in the course and found externally.

<br> Frequent itemset mining is the optimal choice because the dataset is organized as transactions with many items. This is extended by association rule mining, which checks the strength of the relationships between the products. However, These methods treat the transactions as unordered sets, which implies they dont capture temporal ordering of the purchases.

<br> Due to this limitation, it is wise to use the sequential pattern mining as an external technique. By utilizing the time information, sequential methods may reveal patterns that are not visible to association rules.

<br> The dataset has already been cleaned and arranged into transactions, making the project feasible. Standard Python libraries support FP-Growth and association rule generation, and sequential pattern mining can be used by utilizing existing tools. Some risks include sensitivity to supporth thresholds and the computational cost for sequential method. These can be dealt with and controlled with careful experimentation and parameter tuning.

<br>This problem is non-trivial because frequent itemset methods ignore ordering, which may cause missing significant temporal relationships in the data.

In [ ]:
# Feasibility check
try:
    from mlxtend.frequent_patterns import fpgrowth
    print("FP-Growth is ready to use.")
except:
    print("mlxtend needs to be installed.")

FP-Growth is ready to use.


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

# Methodological Planning
The project will start with frequent itemset mining using FP-Growth, and Apriori will be used as a baseline method for comparison. These methods will find product combinations that are often used in transactions. FP-Growth would be preffered because it is more efficient than Apriori for larger datasets, as it does not generate a huge number of candidate itemsets.

<br> The frequent itemsets will then be used to generate association rules. To assess the quality of these rules, Support, Confidence, and Lift will be used. These metrics will help figure out if the often occuring rules are meaningful and informative.

<br> Lastly, sequential pattern mining is used to examine the order of purchases over time. This step will help in figuring out whether utilizing the temporal data reveals more structure in customer purchasing behavior.

<br> Inital testing will focus on confirming that the dataset is correctly formatted for each method and that the libraries function as intended. Before scaling up to the entire dataset, subsets of the data will be used to check computational feasibility.